File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int buy = INT_MAX;
        int sell = INT_MIN;
        int lcurr = 0;
        int rcurr = 0;
        int ret = 0;

        for (int i = 0; i < prices.size(); i++) {
            int price = prices[i];
            if (price <= buy) {
                lcurr = i;
                buy = price;
            } else {
                rcurr = i;
                sell = price;
            }

            if (lcurr < rcurr) {
                ret = max(ret, sell - buy);
            } 
        }

        return ret;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        lo: int = 100000000
        hi: int = -1
        ret: int = 0
        for i in range(len(prices)):
            if prices[i] < lo:
                lo = prices[i]
                hi = -1
            elif prices[i] > hi:
                hi = prices[i]
            ret = max(ret, hi-lo)
        return ret
            
            